### PR TITLE
Add GCM support for EVP_CTRL_GCM_IV_GEN/EVP_CTRL_GCM_SET_IV_INV to providers

### PR DESCRIPTION
**Pull request by slontis**
##### Checklist

- [x] documentation is added or updated
- [ ] tests are added or updated

### CODE REVIEW

File: providers/common/ciphers/cipher_gcm.c [hunk 1]
```diff
@@ -138,7 +186,13 @@ int gcm_get_ctx_params(void *vctx, OSSL_PARAM params[])
             return 0;
         }
     }
-
+    p = OSSL_PARAM_locate(params, OSSL_CIPHER_PARAM_GCM_GET_IV_GEN);
+    if (p != NULL) {
+        if (p->data == NULL
+            || p->data_type != OSSL_PARAM_OCTET_STRING
+            || !getivgen(ctx, p->data, p->data_size))
```
Comment (slontis) [inline providers/common/ciphers/cipher_gcm.c]: @levitte - is there a better way to do this?

Comment (levitte) [reply providers/common/ciphers/cipher_gcm.c]: Not that I can think of for the moment...

**Comment by petrovr** [comment]
Tested manually with following command:
$ ./ssh -t -o Ciphers=aes128-gcm@openssh.com localhost tty
```
/* PKIX-SSH *\
\* non-fips */
/dev/pts/3
Connection to localhost closed.
```
i.e. pass .


This modification does not provide correction for one of questions raised 9524 - 
could replace EVP_CTRL_GCM_SET_IV_FIXED with EVP_CTRL_AEAD_SET_IV_FIXED in evp_enc.c ?

Does CCM support SET_IV_FIXED?
$ grep -n EVP_CTRL_'GCM\|AEAD' crypto/evp/evp_enc.c include/openssl/evp.h:
<pre>
crypto/evp/evp_enc.c:1099:    case EVP_CTRL_AEAD_SET_IVLEN:
crypto/evp/evp_enc.c:1104:    case <b>EVP_CTRL_GCM_SET_IV_FIXED</b>:
crypto/evp/evp_enc.c:1106:            OSSL_PARAM_construct_octet_string(OSSL_CIPHER_PARAM_AEAD_TLS1_IV_FIXED,
crypto/evp/evp_enc.c:1109:    case EVP_CTRL_GCM_IV_GEN:
crypto/evp/evp_enc.c:1117:    case EVP_CTRL_GCM_SET_IV_INV:
crypto/evp/evp_enc.c:1132:    case EVP_CTRL_AEAD_GET_TAG:
crypto/evp/evp_enc.c:1134:    case EVP_CTRL_AEAD_SET_TAG:
crypto/evp/evp_enc.c:1135:        params[0] = OSSL_PARAM_construct_octet_string(OSSL_CIPHER_PARAM_AEAD_TAG,
crypto/evp/evp_enc.c:1138:    case EVP_CTRL_AEAD_TLS1_AAD:
crypto/evp/evp_enc.c:1141:            OSSL_PARAM_construct_octet_string(OSSL_CIPHER_PARAM_AEAD_TLS1_AAD,
crypto/evp/evp_enc.c:1147:            OSSL_PARAM_construct_size_t(OSSL_CIPHER_PARAM_AEAD_TLS1_AAD_PAD, &sz);
include/openssl/evp.h:291:# define         EVP_CIPH_FLAG_AEAD_CIPHER       0x200000
include/openssl/evp.h:316:# define         EVP_CTRL_AEAD_SET_IVLEN         0x9
include/openssl/evp.h:317:# define         EVP_CTRL_AEAD_GET_TAG           0x10
include/openssl/evp.h:318:# define         EVP_CTRL_AEAD_SET_TAG           0x11
include/openssl/evp.h:319:# define         EVP_CTRL_AEAD_SET_IV_FIXED      0x12
include/openssl/evp.h:320:# define         EVP_CTRL_GCM_SET_IVLEN          EVP_CTRL_AEAD_SET_IVLEN
include/openssl/evp.h:321:# define         EVP_CTRL_GCM_GET_TAG            EVP_CTRL_AEAD_GET_TAG
include/openssl/evp.h:322:# define         EVP_CTRL_GCM_SET_TAG            EVP_CTRL_AEAD_SET_TAG
include/openssl/evp.h:323:# define         <b>EVP_CTRL_GCM_SET_IV_FIXED       <i>EVP_CTRL_AEAD_SET_IV_FIXED</i></b>
include/openssl/evp.h:324:# define         EVP_CTRL_GCM_IV_GEN             0x13
include/openssl/evp.h:325:# define         EVP_CTRL_CCM_SET_IVLEN          EVP_CTRL_AEAD_SET_IVLEN
include/openssl/evp.h:326:# define         EVP_CTRL_CCM_GET_TAG            EVP_CTRL_AEAD_GET_TAG
include/openssl/evp.h:327:# define         EVP_CTRL_CCM_SET_TAG            EVP_CTRL_AEAD_SET_TAG
include/openssl/evp.h:328:# define         <b>EVP_CTRL_CCM_SET_IV_FIXED       <i>EVP_CTRL_AEAD_SET_IV_FIXED</i></b>
include/openssl/evp.h:332: * AEAD cipher deduces payload length and returns number of bytes required to
include/openssl/evp.h:336:# define         EVP_CTRL_AEAD_TLS1_AAD          0x16
include/openssl/evp.h:337:/* Used by composite AEAD ciphers, no-op in GCM, CCM... */
include/openssl/evp.h:338:# define         EVP_CTRL_AEAD_SET_MAC_KEY       0x17
include/openssl/evp.h:340:# define         EVP_CTRL_GCM_SET_IV_INV         0x18
include/openssl/evp.h:384:# define         EVP_AEAD_TLS1_AAD_LEN           13
</pre>

**Comment by slontis** [comment]
EVP_CTRL_GCM_SET_IV_FIXED & EVP_CTRL_AEAD_SET_IV_FIXED are defined as the same thing, so it makes no difference which one is used.
SET_IV_FIXED is supported by CCM.

**Comment by slontis** [comment]
ping

**Comment by slontis** [comment]
ping: rebased

**Comment by slontis** [comment]
ping: rebased again
@levitte this will conflict with your PR where you move files around

**Comment by petrovr** [comment]
Author refuse to replace EVP_CTRL_GCM_SET_IV_FIXED with EVP_CTRL_AEAD_SET_IV_FIXED - improved readability.
Team refuse to approve restore of lost functionality.

So issue is in deadlock state.

File: doc/man7/provider-cipher.pod [hunk 2]
```diff
@@ -349,6 +349,16 @@ by AES SIV ciphers which disallow multiple operations by default.
...
+=item "gcm_iv_gen" (B<OSSL_CIPHER_PARAM_GCM_GET_IV_GEN>) <octet string>
+
+Gets the GCM invocation field generated for encryption. Can only be called
+after "tlsivfixed" is set.
+
+=item "gcm_iv_inv" (B<OSSL_CIPHER_PARAM_GCM_SET_IV_INV>) <octet string>
+
+Sets the GCM invocation field used for decryption. Can only be called
+after "tlsivfixed" is set.
+
```
Comment (levitte) [inline doc/man7/provider-cipher.pod]: If gcm and ccm parameters of this sort are really AEAD parameters, why do we insist on using "gcm" in the params?  Also, if this is meant specifically for TLS use, why not name these parameters in the same manner as the other TLS specific params, such as "tlsivgen" and "tlsivinv"?  If not meant specifically for TLS use, why not "ivgen" and "ivinv"?

File: crypto/evp/evp_enc.c [hunk 3]
```diff
@@ -1129,11 +1129,26 @@ int EVP_CIPHER_CTX_ctrl(EVP_CIPHER_CTX *ctx, int type, int arg, void *ptr)
...
+            OSSL_PARAM_construct_octet_string(OSSL_CIPHER_PARAM_GCM_GET_IV_GEN,
+                                              ptr, sz);
+        break;
+    case EVP_CTRL_GCM_SET_IV_INV:
+        if (arg < 0)
+            return 0;
+        params[0] =
+            OSSL_PARAM_construct_octet_string(OSSL_CIPHER_PARAM_GCM_SET_IV_INV,
+                                              ptr, sz);
+        break;
```
Comment (levitte) [inline crypto/evp/evp_enc.c]: I see no reason to insist on GCM control names when they are just aliases for AEAD control names...

**Comment by levitte** [comment]
Rebase needed.

**Comment by levitte** [comment]
> Team refuse to approve restore of lost functionality.

Team has quite a few things on its plate and limited resources.  Please do not see delays on *your* favorite topics of interest as refusal.

File: crypto/evp/evp_enc.c (hunk 3 above)
Comment (slontis) [reply crypto/evp/evp_enc.c]: @levitte - EVP_CTRL_GCM_IV_GEN & EVP_CTRL_GCM_SET_IV_INV are existing names without aliases.. I am not sure what you mean.

Comment (levitte) [reply crypto/evp/evp_enc.c]: Oh... Darn, sorry, didn't look closely enough, I had it in my mind that it was the same with those as with iv_fixed. Never mind.

**Comment by slontis** [comment]
> Author refuse to replace EVP_CTRL_GCM_SET_IV_FIXED with EVP_CTRL_AEAD_SET_IV_FIXED - improved readability.

Err this PR has this change..

> Team refuse to approve restore of lost functionality.
> So issue is in deadlock state.

Hopefully this gets resolved a bit quicker now..

File: doc/man7/provider-cipher.pod (hunk 2 above)
Comment (slontis) [reply doc/man7/provider-cipher.pod]: updated names to tlsivgen and tlsivinv (I was just mapping to the control names)

**Comment by slontis** [comment]
ping - rebased due to the provider folders moving.

**Comment by mattcaswell** [comment]
Looks ok to me - but needs a rebase again!!

**Comment by slontis** [comment]
rebased.

**Comment by slontis** [comment]
ping

File: crypto/evp/evp_enc.c [hunk 4]
```diff
@@ -1135,10 +1135,22 @@ int EVP_CIPHER_CTX_ctrl(EVP_CIPHER_CTX *ctx, int type, int arg, void *ptr)
...
+        if (arg < 0)
+            sz = 0; /* special case that uses the iv length */
+        params[0] = OSSL_PARAM_construct_octet_string(
+                        OSSL_CIPHER_PARAM_AEAD_TLS1_GET_IV_GEN, ptr, sz);
+        break;
+    case EVP_CTRL_GCM_SET_IV_INV:
+        if (arg < 0)
+            return 0;
+        params[0] = OSSL_PARAM_construct_octet_string(
+                        OSSL_CIPHER_PARAM_AEAD_TLS1_SET_IV_INV, ptr, sz);
```
Comment (levitte) [inline crypto/evp/evp_enc.c]: At this point, I'm utterly confused.  Are ivgen and ivinv GCM only or not?

Comment (slontis) [reply crypto/evp/evp_enc.c]: As it says in the pod description for these entries - They are only used by GCM..

**Comment by slontis** [comment]
Thanks.. code merged to master.

**Pull request closed without merging**